File: 142.linked-list-cycle-ii.java
/*
 * @lc app=leetcode id=142 lang=java
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null) {
            return null;
        }
        if (head.next == null) {
            return null;
        }
        if (head.next == head) {
            return head;
        }
        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow) {
                ListNode iterA = slow, iterB = head;
                while (iterA != iterB) {
                    iterA = iterA.next;
                    iterB = iterB.next;
                }
                return iterA;
            }
        }
        return null;
    }
}
// @lc code=end

File: 142.linked-list-cycle-ii.cpp
/*
 * @lc app=leetcode id=142 lang=cpp
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (head == nullptr || head->next == nullptr) {
            return nullptr;
        }
        ListNode *fast = head->next->next, *slow = head->next;
        while (fast != slow) {
            if (fast == nullptr) { return nullptr; }
            fast = fast->next;
            if (fast == nullptr) { return nullptr; }
            fast = fast->next;
            slow = slow->next;
        }
        slow = head;
        while (slow != fast) {
            slow = slow->next;
            fast = fast->next;
        }
        return slow;
        
    }
};
// @lc code=end
